e2e/helpers.ts

```diff
@@ -0,0 +1,57 @@
+import { E2EMessage } from './types';
+import { PNG } from 'pngjs';
+import fs from 'node:fs';
+import path from 'path';
+import pixelmatch from 'pixelmatch';
+
+const replacer = (key: string, value: any) => {
+  if (key === 'type' && typeof value !== 'string') return value.displayName;
+  return value;
+};
+
+export const sendToDeviceAndReceive = <R>(message: E2EMessage) =>
+  new Promise<R>((resolve) => {
+    global.client.once('message', (message) => {
+      const parsedMessage: E2EMessage = JSON.parse(message.toString('utf-8'));
+      resolve(parsedMessage as R);
+    });
+    global.client.send(JSON.stringify(message, replacer));
+  });
+
+export const compareImages = (
+  image1: Buffer,
+  image2: Buffer,
+  opts: {
+    width: number;
+    height: number;
+    pixelRatio: number;
+    diffFilePath?: string;
+    renderedFilePath?: string;
+  }
+) => {
+  const referencePng = PNG.sync.read(image1);
+  const responsePng = PNG.sync.read(image2);
+
+  const diffPng = new PNG({
+    height: referencePng.height,
+    width: referencePng.width,
+  });
+
+  const pixelDiff = pixelmatch(
+    referencePng.data,
+    responsePng.data,
+    diffPng.data,
+    opts.width * opts.pixelRatio,
+    opts.height * opts.pixelRatio,
+    {
+      diffColor: [95, 0, 160],
```

Comment:
Can we get a comment here or something? They look like magic numbers.

Reply:
Note for @bohdanprog: they are indeed, it's just random RGB color.